### PR TITLE
Datum Iterators

### PR DESCRIPTION
**Pull request by adelelopez**
DatumFactory requires datums to be acquired by index. This becomes very impractical for implementing #3623.

This POC is a design for an alternative, called DatumIterator. This works like a standard stateful iterator, and has some nice compositionality properties.

On the other hand, it will require some rethinking of how datum chunks are handled, since those are currently accessed by going from low to high. Using the same iterator and stopping at the end of each chunk should work, but it will be a non-trivial change.

### CODE REVIEW

File: src/server/worker/datum_iterator.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		result += datumIterator.Len()
+	}
+	return result
+}
+
+func (d *unionDatumIterator) Next() bool {
+	if len(d.inputs) == 0 {
+		return false
+	}
+	inputHasNext := d.inputs[d.unionIdx].Next()
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: Is this correct? I don't think this would work as intended. It looks like it returns false when the first inner Next returns true, and does not seem like it would handle the case when the second inner Next returns false (0 inputs) but is followed by additional datum iterators. If I am understanding this correctly, I think it might make sense to refactor this to be recursive. So, you could have something like this: 
```
if d.unionIdx >= len(d.inputs) {
  return false
}
if !d.inputs[d.unionIdx].Next() {
  d.unionIdx++
  return d.Next()
}
return true
```
Also, it might make sense to rename `d.inputs` to `d.iterators` or something to indicate they are iterators since it is a little confusing to overload the word inputs.

File: src/server/worker/datum_iterator.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	}
+	result := d.inputs[0].Len()
+	for i := 1; i < len(d.inputs); i++ {
+		result *= d.inputs[i].Len()
+	}
+	return result
+}
+
+func (d *crossDatumIterator) Next() bool {
+	// we need to make sure each of the iterators has been started (i.e is not at -1)
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: I think there should be a check here similar to the union iterator for if the input length is zero.

File: src/server/worker/datum_iterator.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	hasNext := true
+	if !d.started {
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: I think it might be cleaner to have this handled in the newCrossDatumIterator function and set inputs as empty if it is not a valid cross (relying on the check previously mentioned for correct handling of the Next/Datum pattern).

Comment (brycemcanally) [reply src/server/worker/datum_iterator.go]: Handling the first Next call would just be:
```
if d.first {
  d.first = false
  return true
}
```
`d.first` would be set to true in the `Reset` function.

File: src/server/worker/datum_iterator.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	// this is the normal case
+	for _, input := range d.inputs {
+		// if we're at the end of the "row"
+		if !input.Next() {
+			// we reset the "row"
+			input.Reset()
+
+			if !input.Next() {
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: I'm not sure if we want this check. It seems like it would hide a bug.

File: src/server/worker/datum_iterator.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+				// this should never happen, since the started case should handle this
+				return false
+			}
+			// after resetting this "row", start iterating through the next "row"
+		} else {
+			return true
+		}
+	}
+	return true
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: This should be false right? When we get to the end (the last input gets reset), we are done right?

**Comment by jdoliner** [comment]
`GlobFileStream` doesn't take a range though and I think that would be pretty impossible to implement efficiently. I think chunking by hash is the only answer that's really going to work, since it has the nice property that the chunking remains the same regardless of glob pattern or join operator. I think this is also the route toward not having to store everything in memory, since you can operate on chunks of data by passing those hashes through to the file system layer.

File: src/server/worker/datum_iterator.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	sort.Slice(result.inputs, func(i, j int) bool {
+		// We sort by descending size first because it can boost performance to
+		// process the biggest datums first.
+		if result.inputs[i].FileInfo.SizeBytes != result.inputs[j].FileInfo.SizeBytes {
+			return result.inputs[i].FileInfo.SizeBytes > result.inputs[j].FileInfo.SizeBytes
+		}
+		return result.inputs[i].FileInfo.File.Path < result.inputs[j].FileInfo.File.Path
+	})
+	// make sure it gets initialized properly
+	result.Reset()
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: Just a nitpick, but it might matter in the future. This should probably be executed immediately after creating the result variable (maybe in a defer) because you can return an iterator without initializing for the input.Commit == "" case. This is not a bug now, but could be later with some changes.

File: src/server/worker/datum_iterator.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+func newUnionDatumIterator(pachClient *client.APIClient, union []*pps.Input) (DatumIterator, error) {
+	result := &unionDatumIterator{}
+	for _, input := range union {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.iterators = append(result.iterators, datumIterator)
+	}
+	result.Reset()
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: Same as above, but I think this would have to be deferred. Not really a big deal, but it might be be a good idea to future proof.

File: src/server/worker/datum_iterator.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+	result.inputs = append(
+		result.inputs,
+		&Input{
+			FileInfo: fileInfo,
+			Name:     input.Name,
+			Branch:   input.Branch,
+			GitURL:   input.URL,
+		},
+	)
+	result.Reset()
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: Same as above.

File: src/server/worker/datum_iterator.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+func newCrossDatumIterator(pachClient *client.APIClient, cross []*pps.Input) (DatumIterator, error) {
+	result := &crossDatumIterator{}
+	for _, input := range cross {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		// start the iterator, and make sure it isn't empty
+		if !datumIterator.Next() {
+			return &crossDatumIterator{}, nil
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: I think there is still an issue here without a length check in Next. The first Next call will return true with no inputs because started will be false. I think it might make sense to have the Reset function actually check for a Next in each input and empty the inputs if there is not. So, you could just defer Reset similar to what I mentioned above and remove this check. Then, either add a len(d.inputs) check to Next or set started to true for this edge case.

File: src/server/worker/datum_iterator.go [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+	if !d.started {
+		d.started = true
+		return true
+	}
+	for _, input := range d.inputs {
+		// if we're at the end of the "row"
+		if !input.Next() {
+			// we reset the "row"
+			input.Reset()
+			// and start back it up
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: wording lol.

File: src/server/worker/datum_iterator.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+	}
+	return true
+}
+
+func (d *unionDatumIterator) Datum() []*Input {
+	return d.iterators[d.unionIdx].Datum()
+}
+
+type crossDatumIterator struct {
+	inputs  []DatumIterator
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: It would probably be more consistent to also change this variable name to `iterators`.

File: src/server/worker/datum_iterator_test.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+	tu "github.com/pachyderm/pachyderm/src/server/pkg/testutil"
+)
+
+func TestDatumIterators(t *testing.T) {
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator_test.go]: I think it would be a good idea to expand on this test a bit more. Right now I think it just checks for duplicates and that the number of datums read match the length. It seems to make more sense to me that we should have a hard coded, expected result (list of datums) for each of the types so that we can confirm the order and the actual datums produced to validate that the iterators work as expected.

File: src/server/worker/datum_iterator.go [hunk 13]
```diff
@@ -137,41 +137,44 @@ func (d *unionDatumIterator) Datum() []*Input {
...
+		result.iterators = append(result.iterators, datumIterator)
 	}
 	return result, nil
 }
 
 func (d *crossDatumIterator) Reset() {
-	for _, input := range d.inputs {
-		input.Reset()
-		input.Next()
+	inhabited := true
```
Comment (brycemcanally) [inline src/server/worker/datum_iterator.go]: I think we are going to run into a problem when using this iterator in a composition. I believe `Reset` would get called twice (once when we create the cross iterator, and once when we create the iterator that has the cross within it) which would result with started being set to false after the second call when the cross should produce no datums. The first call will set the iterators to nil, and the second call would never set inhabited to false (therefore making started false). If think something like this should work:
```
for _, iterator := range d.iterators {
	iterator.Reset()
	if !iterator.Next() {
		d.iterators = nil
	}
}
d.started = len(d.iterators) == 0
```

Comment (brycemcanally) [reply src/server/worker/datum_iterator.go]: Also, the updated test looks good. I would just add a test for this particular case (cross that produces zero datums embedded in a union or cross).
